Ensuring initialization of `CaptionUserPreferencesTestingModeToken` before setting caption mode

cola119 opened the pull request:
#### 9948a5b7f3736753adf75857450b9a518806f501
<pre>
Ensuring initialization of `CaptionUserPreferencesTestingModeToken` before setting caption mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=262680">https://bugs.webkit.org/show_bug.cgi?id=262680</a>

Reviewed by Eric Carlson.

The `captionDisplayMode` option was introduced in
<a href="https://bugs.webkit.org/show_bug.cgi?id=261460">https://bugs.webkit.org/show_bug.cgi?id=261460</a> (PR:
<a href="https://github.com/WebKit/WebKit/pull/17689).">https://github.com/WebKit/WebKit/pull/17689).</a> However,
when I attempted to import WPT for webvtt/parsing with
`captionDisplayMode` setting enabled, I found that the
tests were failing intermittently on ios-wk2 (example:
<a href="https://github.com/WebKit/WebKit/pull/17852#issuecomment-1723066317).">https://github.com/WebKit/WebKit/pull/17852#issuecomment-1723066317).</a>

The flakiness of these tests appears to be due to the fact
that the `CaptionUserPreferencesTestingModeToken` instance
may not always be initialized before the caption mode is
set (<a href="https://github.com/WebKit/WebKit/blob/659f85098cf4322282a775cbd51eb58a4ed34022/Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp#L583).">https://github.com/WebKit/WebKit/blob/659f85098cf4322282a775cbd51eb58a4ed34022/Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp#L583).</a>
This instance is usually initialized on the `Internals`
class&apos;s constructor during the WebKitTestRunner setup
(<a href="https://github.com/WebKit/WebKit/blob/659f85098cf4322282a775cbd51eb58a4ed34022/Tools/WebKitTestRunner/InjectedBundle/InjectedBundlePage.cpp#L888).">https://github.com/WebKit/WebKit/blob/659f85098cf4322282a775cbd51eb58a4ed34022/Tools/WebKitTestRunner/InjectedBundle/InjectedBundlePage.cpp#L888).</a>
In most cases, the `CaptionUserPreferencesTestingModeToken`
instance is created in the `Internals` class constructor
before `WKBundlePageSetCaptionDisplayMode` is called, but
there are instances where it is not, particularly when tests
are run concurrently. This leads to the tests failing.

To address this issue, the WebKitTestRunner needs to ensure
that the `CaptionUserPreferencesTestingModeToken` instance
is always initialized before `WKBundlePageSetCaptionDisplayMode`
is called, thus eliminating the flakiness of the tests.

* Source/WebKit/Shared/API/APICaptionUserPreferencesTestingModeToken.h: Added.
(API::CaptionUserPreferencesTestingModeToken::create):
(API::CaptionUserPreferencesTestingModeToken::CaptionUserPreferencesTestingModeToken):
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/API/c/WKBase.h:
* Source/WebKit/UIProcess/API/C/WKAPICast.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundlePageSetCaptionDisplayMode):
(WKBundlePageCreateCaptionUserPreferencesTestingModeToken):
(WKBundleSetCaptionDisplayMode): Deleted.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePagePrivate.h:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::beginTesting):
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.h:

Canonical link: <a href="https://commits.webkit.org/269455@main">https://commits.webkit.org/269455@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ecd2ea3a75643258f471b87c3888c7a0a9072a4e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22337 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22535 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23413 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24238 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20672 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22588 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26798 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22867 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21683 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22573 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22270 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19389 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25093 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19320 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20244 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26493 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20337 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20475 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24353 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/21007 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17799 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20462 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5385 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24695 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20999 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
```diff
@@ -849,7 +849,18 @@ void WKBundlePageClearApplicationCache(WKBundlePageRef page)
 }
 
-void WKBundleSetCaptionDisplayMode(WKBundlePageRef page, WKStringRef mode)
+void WKBundlePageCreateCaptionUserPreferencesTestingModeToken(WKBundlePageRef page, std::unique_ptr<WebCore::CaptionUserPreferencesTestingModeToken>* token)
```
`std::unique_ptr<T>*` looks odd. You need to reseat the unique pointer, so passing as an lvalue reference makes your intent much more clear. And then use as such:

```cpp
void WKBundlePageCreateCaptionUserPreferencesTestingModeToken(WKBundlePageRef page, std::unique_ptr<WebCore::CaptionUserPreferencesTestingModeToken>& token)
{
    // ...
    token = captionPreferences.createTestingModeToken().moveToUniquePtr();
    // ...
}
```

emw-apple commented on Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.h:
```diff
@@ -27,6 +27,7 @@
 #define WKBundlePage_h
 
 #include <JavaScriptCore/JavaScript.h>
+#include <WebCore/CaptionUserPreferences.h>
```
On Apple platforms, we treat WebCore as an implementation detail and do not support _clients_ of WebKit including WebCore headers. This means that any headers which go into the SDK (these are marked "public" and "private" headers in Xcode) need to avoid including WebCore, and I think it's why your iOS/tvOS/watchOS builds are failing.

Instead of directly including, you should try adding a line for `CaptionUserPreferences.h` to Source/WebKit/MigratedHeaders-input.xcfilelist and Source/WebKit/MigratedHeaders-output.xcfilelist. There's a build script that copies these headers into WebKit proper, and fixes them up so that they are importable without a bunch of WebCore types.

emw-apple replied:
and once you've added the line to the MigratedHeaders xcfilelists, you'd change this include statement to be
```suggestion
#include <WebKit/CaptionUserPreferences.h>
```

cola119 replied:
Thank you for your advice. As per your suggestion, I have tried to add a line for `CaptionUserPreferences.h` to `MigrateHeaders-input/output.xcfilelist`. Then I changed the include statement. However, after making these changes, I encountered a series of errors related to the `namespace` and `using` keywords not being recognized. It seems like the C++ headers are being compiled as C code. I'm not sure how to proceed from here and would appreciate any further guidance.

emw-apple replied:
Ah, yes, this is C API.

Maybe you can find a simpler way to do this, but I think you could create an opaque type for CaptureUserPreferencesTestingModeToken in `WKBase.h`, add a mapping in `WKAPICast.h`, and a C API to create an opaque token. Then the test runner wouldn't need to know about the (WebCore) C++ types at all.

cola119 replied:
I've implemented a type mapping for CaptureUserPreferencesTestingModeToken so that C API can handle it without knowing the implementation details in WebCore. Could you review it again? Thank you.

cola119 committed:
Ensuring initialization of `CaptionUserPreferencesTestingModeToken` before setting caption mode
https://bugs.webkit.org/show_bug.cgi?id=262680

Reviewed by Eric Carlson.

The `captionDisplayMode` option was introduced in
https://bugs.webkit.org/show_bug.cgi?id=261460 (PR:
https://github.com/WebKit/WebKit/pull/17689). However,
when I attempted to import WPT for webvtt/parsing with
`captionDisplayMode` setting enabled, I found that the
tests were failing intermittently on ios-wk2 (example:
https://github.com/WebKit/WebKit/pull/17852#issuecomment-1723066317).

The flakiness of these tests appears to be due to the fact
that the `CaptionUserPreferencesTestingModeToken` instance
may not always be initialized before the caption mode is
set (https://github.com/WebKit/WebKit/blob/659f85098cf4322282a775cbd51eb58a4ed34022/Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp#L583).
This instance is usually initialized on the `Internals`
class's constructor during the WebKitTestRunner setup
(https://github.com/WebKit/WebKit/blob/659f85098cf4322282a775cbd51eb58a4ed34022/Tools/WebKitTestRunner/InjectedBundle/InjectedBundlePage.cpp#L888).
In most cases, the `CaptionUserPreferencesTestingModeToken`
instance is created in the `Internals` class constructor
before `WKBundlePageSetCaptionDisplayMode` is called, but
there are instances where it is not, particularly when tests
are run concurrently. This leads to the tests failing.

To address this issue, the WebKitTestRunner needs to ensure
that the `CaptionUserPreferencesTestingModeToken` instance
is always initialized before `WKBundlePageSetCaptionDisplayMode`
is called, thus eliminating the flakiness of the tests.

* Source/WebKit/Shared/API/APICaptionUserPreferencesTestingModeToken.h: Added.
(API::CaptionUserPreferencesTestingModeToken::create):
(API::CaptionUserPreferencesTestingModeToken::CaptionUserPreferencesTestingModeToken):
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/API/c/WKBase.h:
* Source/WebKit/UIProcess/API/C/WKAPICast.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
(WKBundlePageSetCaptionDisplayMode):
(WKBundlePageCreateCaptionUserPreferencesTestingModeToken):
(WKBundleSetCaptionDisplayMode): Deleted.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePagePrivate.h:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::beginTesting):
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.h:

Canonical link: https://commits.webkit.org/269455@main
